src/logic/map_objects/tribes/ship.cc

```diff
@@ -637,21 +662,14 @@ bool Ship::ship_update_expedition(Game& game, Bob::State& /* state */) {
 		// Look for new nearby port spaces.
 		MapRegion<Area<Coords>> mr(*map, Area<Coords>(position, descr().vision_range()));
 		do {
-			if (map->is_port_space(mr.location()) &&
+			if (suited_as_invasion_portspace(mr.location()) &&
 			    std::find(portspaces.begin(), portspaces.end(), mr.location()) == portspaces.end()) {
-				const Field& field = (*map)[mr.location()];
-				if (field.get_immovable() != nullptr &&
-				    field.get_immovable()->get_owner() == get_owner() &&
-				    game.descriptions().building_index(field.get_immovable()->descr().name()) ==
-				       owner().tribe().port()) {
-					continue;  // Skip if the player already has a port there
-				}
-
-				found_new_target = true;
 				portspaces.push_back(mr.location());
-				remember_detected_portspace(mr.location());
-				send_message(game, _("Port Space"), _("Port Space Found"),
-				             _("A warship found a new port build space."), descr().icon_filename());
+				if (remember_detected_portspace(mr.location())) {
+					found_new_target = true;
+					send_message(game, _("Port Space"), _("Port Space Found"),
+					             _("A warship found a new port build space."), descr().icon_filename());
+				}
```

Comment:
This way it prevents finding (and thus attacking) any previously seen portspace, not just the ones that were in sight in the previous step.

I meant like it is for civilian expedition ships in lines 792-834. Though that has a bug too: if it finds a new portspace next to the previous one, then the icon is still shown in the previous position. So it's not clear for the user that it's a different port space.

Ideally still seen but ignored port spaces should be shown with some transparency, like when an expedition sail away.